src/services/pcn-packetcapture/src/Packetcapture_dp_egress.c

```diff
@@ -0,0 +1,163 @@
+  __be64 src : 48;
+  __be16 proto;
+} __attribute__((packed));
+
+
+/*
+ * BPF map where a single element, working mode [ON | OFF]
+ */
+BPF_ARRAY(working, uint8_t, 1);
+
+/*
+ * BPF map where a single element, the filters map
+ */
+BPF_ARRAY(filters_map, struct filters_table, 1);
+
+
+static __always_inline int handle_rx(struct CTXTYPE *ctx, struct pkt_metadata *md) {
+
+  unsigned int key = 0;
+  uint8_t *status = working.lookup(&key);
+  if(!status){
+      return RX_DROP;
+  }
+  
+  if(*status == OFF){
+      return RX_OK;
+  }
+
+  /* Parsing L2 */
+  void *data = (void *)(long)ctx->data;
+  void *data_end = (void *)(long)ctx->data_end;
+  struct eth_hdr *ethernet = data;
+  if (data + sizeof(*ethernet) > data_end)
+    return RX_DROP;
+
+  uint16_t ether_type = ethernet->proto;
+
+  if (ctx->vlan_present) {
+    ether_type = ctx->vlan_proto;
+  }
+
+  /* Parsing L3 */
+  struct iphdr *ip = NULL;
+  struct tcphdr *tcp = NULL;
+  struct udphdr *udp = NULL;
+
+  if (ether_type == bpf_htons(ETH_P_IP)) {
+    ip = data + sizeof(*ethernet);  
+    if (data + sizeof(*ethernet) + sizeof(*ip) > data_end)
+      return RX_DROP;   
+    if((int)ip->version != 4){
+      return RX_OK;
+    }
+
+    struct filters_table *filters_tab;
+    filters_tab = filters_map.lookup(&key);
+    if (filters_tab == NULL) {
+      return RX_DROP;
+    }
+    
+    /* Souce Ip filter */
+    if((filters_tab->network_filter_src_flag == true) && ((filters_tab->netmask_filter_src & ntohl(ip->saddr)) != filters_tab->network_filter_src)){
+      return RX_OK;
+    }
+      
+    /* Destination Ip filter */
+    if((filters_tab->network_filter_dst_flag == true) && ((filters_tab->netmask_filter_dst & ntohl(ip->daddr)) != filters_tab->network_filter_dst)){
+      return RX_OK;
+    }
+
+    /* l4proto Filter */
+    if(filters_tab->l4proto_flag == true){
+      if((filters_tab->l4proto_filter == 1) && (ip->protocol != IPPROTO_TCP)){         /* tcp only */
+        return RX_OK;
+      }else if((filters_tab->l4proto_filter == 2) && (ip->protocol != IPPROTO_UDP)){   /* udp only */
+        return RX_OK;
+      }
+    }
+
+    unsigned int iphdr_size = (unsigned int) ip->ihl * 4;
+    if (ip->protocol == IPPROTO_TCP) {
+      tcp = data + sizeof(*ethernet) + iphdr_size;
+      if (data + sizeof(*ethernet) + iphdr_size + sizeof(*tcp) > data_end)
+        return RX_DROP;
+
+      /* src port filter */
+      if((filters_tab->src_port_flag == true) && (filters_tab->src_port_filter != ntohs(tcp->source))){
+        return RX_OK;
+      }
+
+      /* dst port filter */
+      if((filters_tab->dst_port_flag == true) && (filters_tab->dst_port_filter != ntohs(tcp->dest))){
+        return RX_OK;
+      }
+
+    } else if (ip->protocol == IPPROTO_UDP) {
+      udp = data + sizeof(*ethernet) + iphdr_size;
+      if (data + sizeof(*ethernet) + iphdr_size + sizeof(*udp) > data_end)
+        return RX_DROP;
+    }
+
+  }
+  u16 reason = 1;
+  return pcn_pkt_controller(ctx, md, reason);
```

Comment:
This introduces a big performance penalty, as all the traffic is sent to controller and then back here.
Please add a big "todo" here, suggesting an alternative implementation and telling people why you were forced to stay with this behavior.